xmtp_mls/src/storage/sql_key_store.rs

```diff
@@ -1,87 +1,1052 @@
+        let key = epoch_key_pairs_id(group_id, epoch, leaf_index)?;
+        let value = serde_json::to_vec(key_pairs)?;
+        log::debug!("Writing encryption epoch key pairs");
+        #[cfg(feature = "test-utils")]
+        {
+            log::debug!("  key: {}", hex::encode(&key));
+            log::debug!("  value: {}", hex::encode(&value));
         }
-        db_deserialize(&entry_option.unwrap().value_bytes).ok()
-    }
-
-    /// Delete a value stored for ID `k`.
-    ///
-    /// Interface is unclear on expected behavior when item is already deleted -
-    /// we choose to not surface an error if this is the case.
-    fn delete<V: MlsEntity>(&self, k: &[u8]) -> Result<(), Self::Error> {
-        let conn: &dyn Delete<StoredKeyStoreEntry, Key = Vec<u8>> = self.conn();
-        let num_deleted = conn.delete(k.to_vec())?;
-        if num_deleted == 0 {
-            debug!("No entry to delete for key {:?}", k);
+
+        self.write::<CURRENT_VERSION>(EPOCH_KEY_PAIRS_LABEL, &key, &value)
+    }
+
+    fn encryption_epoch_key_pairs<
+        GroupId: traits::GroupId<CURRENT_VERSION>,
+        EpochKey: traits::EpochKey<CURRENT_VERSION>,
+        HpkeKeyPair: traits::HpkeKeyPair<CURRENT_VERSION>,
+    >(
+        &self,
+        group_id: &GroupId,
+        epoch: &EpochKey,
+        leaf_index: u32,
+    ) -> Result<Vec<HpkeKeyPair>, Self::Error> {
+        let key = epoch_key_pairs_id(group_id, epoch, leaf_index)?;
+        let storage_key = build_key_from_vec::<CURRENT_VERSION>(EPOCH_KEY_PAIRS_LABEL, key);
+        log::debug!("Reading encryption epoch key pairs");
+
+        match self.read_list(EPOCH_KEY_PAIRS_LABEL, &storage_key) {
+            Ok(data) => {
+                #[cfg(feature = "test-utils")]
+                log::debug!("  value: {}", hex::encode(&data));
+                serde_json::from_slice::<Vec<HpkeKeyPair>>(&data)
+                    .map_err(|e| MemoryStorageError::SerializationError)
+            }
+            Err(e) => {
+                log::error!("Failed to read from storage: {}", e);
+                Err(e)
+            }
         }
+    }
+
+    fn delete_encryption_epoch_key_pairs<
+        GroupId: traits::GroupId<CURRENT_VERSION>,
+        EpochKey: traits::EpochKey<CURRENT_VERSION>,
+    >(
+        &self,
+        group_id: &GroupId,
+        epoch: &EpochKey,
+        leaf_index: u32,
+    ) -> Result<(), Self::Error> {
+        let key = epoch_key_pairs_id(group_id, epoch, leaf_index)?;
+        self.delete::<CURRENT_VERSION>(EPOCH_KEY_PAIRS_LABEL, &key)
+    }
+
+    fn clear_proposal_queue<GroupId: traits::GroupId<CURRENT_VERSION>>(
+        &self,
+        group_id: &GroupId,
+    ) -> Result<(), Self::Error> {
+        // I think this key is incorrect and should be let key = serde_json::to_vec(&(group_id, &proposal_ref))?;
+        let key = build_key::<CURRENT_VERSION, &GroupId>(QUEUED_PROPOSAL_LABEL, group_id);
+        let _ = self.delete::<CURRENT_VERSION>(QUEUED_PROPOSAL_LABEL, &key);
```

Comment:
I think we need the `proposal_ref` here. Was this just over looked? Or is this supposed to be a way to clear all of the proposal queue?

Reply:
This needs a small change in the API I think. I made a PR on OpenMLS to fix this and also add a test for the proposals. https://github.com/openmls/openmls/pull/1579
wdyt?